File: java/app/src/main/java/com/pgssoft/mvvm/services/api/RetrofitService.java
package com.pgssoft.mvvm.services.api;

import com.pgssoft.mvvm.model.api.ApiTable;
import com.pgssoft.mvvm.services.exceptions.ApiRequestFailureError;
import com.pgssoft.mvvm.services.interfaces.ApiService;
import com.pgssoft.mvvm.services.interfaces.ApiServiceCallback;

import java.io.IOException;

import okhttp3.OkHttpClient;
import okhttp3.ResponseBody;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

/**
 * Created by bstokrocki on 29.01.2017.
 */
public class RetrofitService implements ApiService {
    private Retrofit retrofit;
    private ApiInterface apiInterface;

    public RetrofitService() {
        retrofit = new Retrofit.Builder()
                .baseUrl("http://api.nbp.pl/api/")
                .client(new OkHttpClient())
                .addConverterFactory(GsonConverterFactory.create())
                .build();

        apiInterface = retrofit.create(ApiInterface.class);
    }

    @Override
    public void loadRates(final ApiServiceCallback<ApiTable[]> callback) {
        Call<ApiTable[]> call = apiInterface.getRates();
        call.enqueue(new BaseCallback<>(callback));
    }

    private class BaseCallback<T> implements Callback<T> {
        private ApiServiceCallback<T> serviceCallback;

        public BaseCallback(ApiServiceCallback<T> serviceCallback) {
            this.serviceCallback = serviceCallback;
        }

        @Override
        public void onResponse(Call<T> call, Response<T> response) {
            if (response.isSuccessful()) {
                serviceCallback.onSuccess(response.body());
            } else {
                serviceCallback.onFailure(new ApiRequestFailureError(response.code(),
                        prepareErrorMessage(response.errorBody())));
            }
        }

        @Override
        public void onFailure(Call<T> call, Throwable t) {
            serviceCallback.onFailure(t);
        }

        private String prepareErrorMessage(ResponseBody responseBody) {
            try {
                return responseBody.string();
            } catch (IOException e) {
                e.printStackTrace();
                return "";
            }
        }
    }
}

File: kotlin/app/src/main/java/com/pgssoft/kotlinplayground/service/api/RetrofitService.kt
package com.pgssoft.kotlinplayground.service.api

import com.pgssoft.kotlinplayground.model.api.ApiTable
import com.pgssoft.kotlinplayground.service.exception.ApiRequestFailureError
import com.pgssoft.kotlinplayground.service.iface.ApiService
import com.pgssoft.kotlinplayground.service.iface.ApiServiceCallback
import okhttp3.OkHttpClient
import okhttp3.ResponseBody
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import java.io.IOException

/**
 * Created by bstokrocki on 29.01.2017.
 */
class RetrofitService : ApiService {
    val apiInterface: ApiInterface

    init {
        val retrofit = Retrofit.Builder()
                .baseUrl("http://api.nbp.pl/api/")
                .client(OkHttpClient())
                .addConverterFactory(GsonConverterFactory.create())
                .build()

        apiInterface = retrofit.create(ApiInterface::class.java)
    }

    override fun loadRates(callback: ApiServiceCallback<Array<ApiTable>>) {
        val call = apiInterface.getRates()
        call.enqueue(BaseCallback(callback))
    }

    private inner class BaseCallback<T>(private val serviceCallback: ApiServiceCallback<T>) : Callback<T> {

        override fun onResponse(call: Call<T>, response: Response<T>) {
            if (response.isSuccessful) {
                serviceCallback.onSuccess(response.body())
            } else {
                serviceCallback.onFailure(ApiRequestFailureError(response.code(),
                        prepareErrorMessage(response.errorBody())))
            }
        }

        override fun onFailure(call: Call<T>, t: Throwable) {
            serviceCallback.onFailure(t)
        }

        private fun prepareErrorMessage(responseBody: ResponseBody): String {
            try {
                return responseBody.string()
            } catch (e: IOException) {
                e.printStackTrace()
                return ""
            }

        }
    }
}
